### PR TITLE
DMTCP_DEBUG_MTCP_RESTART env var: easier debugging.

### PR DESCRIPTION
**Pull request by karya0**

### CODE REVIEW

**Review by gc00 — COMMENTED** [review]
Thanks, @karya0, for this PR.  It will be very valuable for debugging.

Here are some suggestions.  If you're busy, tell me, and I'll directly add these suggestions onto your PR, myself.

1.  Let's choose some more meaningful names.  For the environment variable, perhaps: `DMTCP_DEBUG_MTCP_RESTART `.  (Motivation:  Keep DMTCP environment variables close to each other for grep, alphabetical order, and less namespace polution.)
2. Let's add a comment near where we refer to it:  `// Set this environment variable to debug with GDB inside mtcp_restart.c;` and `//   (Caveat:  not as robust as standard mtcp_restart.)`
3. For the commit message, how about something like: `DMTCP_DEBUG_MTCP_RESTART env var: easier debugging `
4. Let's add a second commit that changes the Makeffile for mtcp_restart to use `-Wl,-Ttext-segment=5555500000 -Tdata=5555560000 ...` (or whatevver the magic incantation was.  I think it's fine to use these addresses in all cases, even if we are not setting the enviornnent variable.  It doesn't hurt the general case, since the position-idependent code can always be imigrated elsewhere.

**Comment by gc00** [comment]
@karya0, Ping.

I think the requests in my comments are simple and easy to apply.  While we both have the spare cycles, let's not leave any orphan PRs.

**Comment by karya0** [comment]
@gc00: Please take a look at it when you get a chance.

**Comment by gc00** [comment]
@karya0, Thanks for addressing the first three points.  Can we address the fourth point?  It should be easy.

Either option A or B:
OPTION A. Let's add a second commit that changes the Makefile for mtcp_restart to use -Wl,-Ttext-segment=5555500000 -Tdata=5555560000 ... (or whatever the magic incantation was).  DMTCP_DEBUG_MTCP_RESTART won't help unless the person also compiles mtcp_restart to a safe address that is unlikely to conflict with the memory regions of the checkponit image.

(I think it's fine to use these addresses in all cases, even if we are not setting the environment variable. It doesn't hurt the general case, since the position-independent code can always be imigrated elsewhere.)

OR:
OPTION B.  If you prefer not to change the default load address for the text and data segments of mtcp_restart, another alternative is to include a comment somewhere near the documentation of the DMTCP_DEBUG_MTCP_RESTART environment variable, explaining to a user that the environment variable is not useful unless they change the load addresses.  (But I think option A is simpler.)

Best,
- Gene

**Comment by karya0** [comment]
@gc00: The load address is already defaulting to a fixed location. Please take a look at Makefile.in.

File: src/mtcp/Makefile.in [hunk 1]
```diff
@@ -88,8 +88,12 @@ default: build
 libs: build
 build: $(targetdir)/bin/$(MTCP_RESTART)
 
+# TODO(kapil): Check if -ttext-segment value is appropriate for x86.
+# The other sections (bss, rodata, etc.) will follow immediately after text.
+DEBUG_FLAGS= -Wl,-Ttext-segment=11100000 #-Wl,-Trodata-segment=445000000
```
Comment (gc00) [inline src/mtcp/Makefile.in:93]: Shouldn't the address be written in hex?  Something like:  `-Wl,-Ttext-segment=0x7000000`
When debugging it's easier if this is a round number in hex.

File: src/mtcp/mtcp_restart.c [hunk 2]
```diff
@@ -246,6 +247,14 @@ main(int argc, char *argv[], char **environ)
     mtcp_abort();
     restart_slow_path();
   } else {
+    // Set this environment variable to debug with GDB inside mtcp_restart.c;
```
Comment (gc00) [inline src/mtcp/mtcp_restart.c:250]: 1.  Is this comment accurate?  I think the comment should say that you need also to `rm bin/mtcp_restart && make` to re-make `bin/mtcp_restart` to load at the new address.  How about the more precise comment (still in 80 char's):
   `// Set env. var. DMTCP_DEBUG_MTCP_RESTART for easy debugging of 'mtcp_restart'. `
2. OPTIONAL COMMENT TO ADD: 
   `// This assumes no address conflict of mtcp_restart (0x111000000) w/ ckpt image.`

**Review by gc00 — APPROVED** [review]
@karya0, Sorry for my delay in reviewing this.

I have two inline comments:
1. -Wl,-Ttest-segment should use hex address
2. improved comment describing that one needs to recompile before using the env. var.

I'm approving this PR in advance for efficiency.  Please take a look at the two inline comments.

File: src/mtcp/Makefile.in (hunk 1 above)
Comment (karya0) [reply src/mtcp/Makefile.in:93]: The address is hex, but without a `0x` prefix. The compiler address-related flags expect hex addresses without a `0x` prefix.

File: src/mtcp/mtcp_restart.c (hunk 2 above)
Comment (karya0) [reply src/mtcp/mtcp_restart.c:250]: We don't need to recompile as the mtcp_restart binary is defaulting to this new address.

**Pull request merged**